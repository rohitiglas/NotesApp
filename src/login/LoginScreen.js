import React, {useState} from 'react';
import {
  Alert,
  TextInput,
  Text,
  View,
  TouchableOpacity,
  StatusBar,
} from 'react-native';
import styles from './LoginStyle';

export const ValidationErrors = {
  FormEmpty: 'Form fields cannot be blank',
  UsernameEmpty: 'Username cannot be blank',
  PasswordEmpty: 'Password cannot be blank',
};
const Login = () => {
  const [userName, setUserName] = useState('');
  const [password, setPassword] = useState('');
  const [validationError, setValidationError] = useState('');

  const onLoginClick = () => {
    if (userName.length === 0 && password.length === 0) {
      setValidationError(ValidationErrors.FormEmpty);
    } else if (userName.length === 0) {
      setValidationError(ValidationErrors.UsernameEmpty);
    } else if (password.length === 0) {
      setValidationError(ValidationErrors.PasswordEmpty);
    } else {
      Alert.alert('Credentials', `${userName} + ${password}`);
    }
  };
  return (
    <View style={styles.mainContainer}>
      {validationError.length !== 0 && (
        <Text testID={'text-error'} style={styles.errorText}>
          {validationError}
        </Text>
      )}

      <View style={styles.loginContainer}>
        <StatusBar barStyle="light-content" />
        <TextInput
          testID={'input-username'}
          value={userName}
          onChangeText={text => setUserName(text)}
          placeholder={'UserName'}
          style={styles.input}
        />
        <TextInput
          testID={'input-password'}
          value={password}
          onChangeText={text => setPassword(text)}
          placeholder={'Password'}
          secureTextEntry={true}
          style={styles.input}
        />
        <TouchableOpacity
          testID={'submit-button'}
          style={styles.buttonContainer}
          onPress={onLoginClick}>
          <Text style={styles.buttonText}>LOGIN</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default Login;
