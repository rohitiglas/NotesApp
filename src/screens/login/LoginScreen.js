import React, {useState} from 'react';
import {
  SafeAreaView,
  Alert,
  Text,
  View,
  TouchableOpacity,
  TouchableNativeFeedback,
  StatusBar,
} from 'react-native';
import styles from './LoginStyle';
import axios from 'axios';
import Input from '../../components/Input';
import Loader from '../../components/Loader';
import {
  _storeData,
  emailValidate,
  resetNavigation,
} from '../../asyncStorage/Local';

export const ValidationErrors = {
  FormEmpty: 'Form fields cannot be blank',
  UsernameEmpty: 'Please enter valid email',
  PasswordEmpty: 'Password cannot be blank',
};
const Login = ({navigation}) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [loader, setLoader] = useState(false);
  const [emailError, setEmailError] = useState('');
  const [passwordError, setPasswordError] = useState('');
  const [validationError, setValidationError] = useState('');

  const onLoginClick = () => {
    if (email.length === 0 && password.length === 0) {
      setValidationError(ValidationErrors.FormEmpty);
    } else if (!emailValidate(email)) {
      setEmailError(ValidationErrors.UsernameEmpty);
    } else if (password.length === 0) {
      setPasswordError(ValidationErrors.PasswordEmpty);
    } else {
      setLoader(true);
      axios
        .post('/api/login', {
          email,
          password,
        })
        .then(async function (response) {
          setLoader(false);
          console.log('Add Notes Response', response);
          Alert.alert('', 'Login Successfully');
          await _storeData(email);
          resetNavigation(navigation, 'Notes');
        })
        .catch(function (error) {
          setLoader(false);
          console.log('Add Notes Error', error);
          Alert.alert('', 'Error adding note.');
        });
    }
  };

  const changeText = (text, type) => {
    if (type === 'email') {
      setEmail(text);
      setEmailError('');
    } else {
      setPassword(text);
      setPasswordError('');
    }
    setValidationError('');
  };
  return (
    <SafeAreaView style={styles.safeAreaViewWrapper}>
      <Loader isLoading={loader} />
      <View style={styles.wrapper}>
        <StatusBar barStyle="light-content" />
        <Input
          style={styles.textInput}
          inputContainerStyle={{borderBottomWidth: 0}}
          testID={'input-username'}
          value={email}
          onChangeText={text => changeText(text, 'email')}
          placeholder={'UserName'}
          placeholderTextColor="#808080"
          error={emailError}
        />
        <Input
          testID={'input-password'}
          value={password}
          onChangeText={text => changeText(text, 'password')}
          placeholder={'Password'}
          secureTextEntry={true}
          style={styles.textInput}
          inputContainerStyle={{borderBottomWidth: 0}}
          placeholderTextColor="#808080"
          error={passwordError}
        />

        {validationError.length !== 0 && (
          <Text testID={'text-error'} style={styles.errorText}>
            {validationError}
          </Text>
        )}
        <TouchableOpacity
          testID={'submit-button'}
          style={styles.buttonContainer}
          onPress={onLoginClick}>
          <Text style={styles.buttonText}>LOGIN</Text>
        </TouchableOpacity>
        <TouchableOpacity
          testID={'forget-password'}
          onPress={() => navigation.navigate('ForgetPassword')}>
          <Text style={styles.forgetText}>Forget Password</Text>
        </TouchableOpacity>

        <View style={{marginTop: 20, flexDirection: 'row'}}>
          <Text
            style={{
              marginLeft: 10,
              color: '#fff',
              fontSize: 16,
              paddingTop: 10,
              paddingBottom: 10,
            }}>
            Not a member of Notes?
          </Text>
          <TouchableNativeFeedback
            onPress={() => {
              navigation.navigate('SignUp');
            }}>
            <Text
              style={{
                marginLeft: 8,
                marginRight: 10,
                color: '#ff2251',
                fontSize: 16,
                fontWeight: '900',
                paddingTop: 10,
                paddingBottom: 10,
              }}>
              Sign up
            </Text>
          </TouchableNativeFeedback>
        </View>
      </View>
    </SafeAreaView>
  );
};

export default Login;
