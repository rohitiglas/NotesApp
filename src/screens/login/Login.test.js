import 'react-native';
import React from 'react';
import {Alert} from 'react-native';
import LoginScreen from './LoginScreen';
import {
  render,
  cleanup,
  fireEvent,
  flushMicrotasksQueue,
} from '@testing-library/react-native';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
let mock = new MockAdapter(axios);
jest.mock('react-native/Libraries/Alert/Alert', () => ({
  alert: jest.fn(),
}));
describe('Login Form', () => {
  const fakeNavigation = {
    navigate: jest.fn(),
  };
  let wrapper;
  beforeEach(() => {
    wrapper = render(<LoginScreen navigation={fakeNavigation} />);
  });
  afterEach(() => {
    cleanup();
    wrapper = null;
  });
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render userId ,Password and Login Button correctly', () => {
    const {getByTestId} = wrapper;
    const userName = getByTestId('input-email');
    const password = getByTestId('input-password');
    const button = getByTestId('submit-button');
    expect(userName).toBeDefined();
    expect(password).toBeDefined();
    expect(button).toBeDefined();
    fireEvent.press(button);
    const error = getByTestId('text-error');
    expect(error.props.children).toBe('Form fields cannot be blank');
  });
  it('should show error when email is empty', () => {
    const {getByTestId} = wrapper;
    const email = getByTestId('input-email');
    const password = getByTestId('input-password');
    const button = getByTestId('submit-button');
    fireEvent.changeText(password, '12345678');
    fireEvent.press(button);
    expect(email.props.error).toBe('Please enter valid email');
  });
  it('should show error when Password is empty', () => {
    const {getByTestId} = wrapper;
    const userName = getByTestId('input-email');
    const password = getByTestId('input-password');
    const button = getByTestId('submit-button');
    fireEvent.changeText(userName, 'Rohit@gmail.com');
    fireEvent.press(button);
    expect(password.props.error).toBe('Password cannot be blank');
  });
  it('should open SignUp Screen on Signup button click', () => {
    const {getByTestId} = wrapper;
    const button = getByTestId('signup-button');
    fireEvent.press(button);
    expect(fakeNavigation.navigate).toBeCalledWith('SignUp');
  });
  it('should open ForgetPassword Screen on Forget button click', () => {
    const {getByTestId} = wrapper;
    const button = getByTestId('forget-password');
    fireEvent.press(button);
    expect(fakeNavigation.navigate).toBeCalledWith('ForgetPassword');
  });
  it('should login api called on submit button', async () => {
    mock.onPost('/api/login').reply(200, {});
    const {getByTestId, queryByTestId} = wrapper;
    const userName = getByTestId('input-email');
    const password = getByTestId('input-password');
    const button = getByTestId('submit-button');
    fireEvent.changeText(userName, 'Rohit@gmail.com');
    fireEvent.changeText(password, '12345678');
    await fireEvent.press(button);
    await flushMicrotasksQueue();
    expect(mock.history.post.length).toBe(1);
    expect(mock.history.post[0].data).toBe(
      JSON.stringify({email: 'Rohit@gmail.com', password: '12345678'}),
    );
    const error = queryByTestId('text-error');
    expect(error).toBe(null);
    const alertSpy = jest.spyOn(Alert, 'alert');
    expect(alertSpy).toHaveBeenCalled();
  });
});
