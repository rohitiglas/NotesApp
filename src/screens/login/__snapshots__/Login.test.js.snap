// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#2c3e50",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="UserName"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-username"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "color": "#2c3e50",
          "height": 50,
          "marginBottom": 20,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input-password"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#2980b6",
          "borderRadius": 20,
          "opacity": 1,
          "paddingVertical": 15,
          "width": "80%",
        }
      }
      testID="submit-button"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        LOGIN
      </Text>
    </View>
  </View>
</View>
`;
